[match] Provisioning Profiles Import and Windows Support

DimitarTachev opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
We are using `Fastlane Match` for managing the code signing in [cloud-based iOS app builds](https://github.com/NativeScript/nativescript-remote-builds) (currently in CircleCI) but we still have a local macOS machine requirement for the `fastlane match` setup while we want to support a full Windows only flow.

### Description
The Pull Request contains two features:

#### 1. Windows support for the current `fastlane match import`.

Details:

When a binary file (in our case the certificates `.cer` one) is opened in **Text Mode on Windows**, the `1a` (`ASCII-26` or `ctrl+Z`) character is assumed as **End Of File** and the provided certificate is not read till the real end. In this way, the `fastlane match import` command is always failing with "_This certificate cannot be imported - the certificate contents did not match with any available on the Developer Portal_" because it is searching for a partial certificate content.

When reading the file in binary mode, the `1a` (`ASCII-26` or `ctrl+Z`) symbol is properly read and the whole file is properly base64 encoded leading to a working certificate import on Windows.

> Additional details:
> http://www.justskins.com/forums/trouble-with-binary-files-105116.html

#### 2. Provisioning Profiles import in the match repository

Details:

This is just the missing part of the `fastlane match import` feature.  It uploads the provisioning profiles along with the certificate and private key into the match repo. 

Note: The new **provisioning profiles path is set as optional** in order to **avoid breaking changes in the command**.

Having this fully working `fastlane match import`, we are now able to work with the Fastlane Match service without any macOS machine requirements. You just need to have the certificate (`.cer`), the private key (`.p12`) and the provisioning profiles (`.mobileprovision or .provisionprofile`) files downloaded on your local Windows machine and run the `fastlane match import` command.

### Testing Steps
We've tested the following steps:
1) Get a clean machine with Windows 10.
2) Install Ruby and OpenSSL
3) Download the certificate (`.cer`), the private key (`.p12`) and the provisioning profiles (`.mobileprovision or .provisionprofile`) files.
4) Upload the signing files using the `fastlane match import` command.
5) Validate that the signing files are properly pushed to the signing repo.
6) Builds the iOS app and publish it to TestFlights using the above-mentioned cloud-based builds and the Fastlane match signing.

DimitarTachev committed:
feat: support match import on Windows

When a binary file is opened in Text Mode on Windows, the 1a (ASCII-26 or ctrl + Z) character is assumed as End Of File and the provided certificate is not read properly. In this way, the `fastlane match import` command is always failing with "This certificate cannot be imported - the certificate contents did not match with any available on the Developer Portal" because it is searching for a partial certificate content.

When reading the file in binary more, the 1a (ASCII-26 or ctrl + Z) is properly read and the whole file is properly base64 encodded leading to a working certificate import on Windows.

More details:
http://www.justskins.com/forums/trouble-with-binary-files-105116.html

DimitarTachev committed:
feat: support provisioning profiles import along with the certificate and private key into the match repo

DimitarTachev committed:
docs: update the match import docs after the Provisioning Profiles support

DimitarTachev commented:
@googlebot I signed it!

DimitarTachev committed:
chore: fix linting errors

joshdholtz commented on match/lib/match/importer.rb:
```diff
@@ -2,23 +2,16 @@
+      # Get and verify cert, p12 and profiles path
+      cert_path = ensure_valid_file_path(cert_path, "Certificate", ".cer")
+      p12_path = ensure_valid_file_path(p12_path, "Private key", ".p12")
+      profile_path = ensure_valid_file_path(profile_path, "Provisioning profile", ".mobileprovision", optional: true)
```
I think we need to also allow `.provisionprofile` here too for Mac profiles 🙃 Are we able to allow multiple types of extensions?

joshdholtz commented on match/lib/match/importer.rb:
```diff
@@ -61,32 +54,53 @@ def import_cert(params, cert_path: nil, p12_path: nil)
+      unless profile_path.nil?
+        FileUtils.mkdir_p(output_dir_profiles)
+        bundle_id = FastlaneCore::ProvisioningProfile.bundle_id(profile_path)
+        dest_profile_path = File.join(output_dir_profiles, "#{cert_type.to_s.capitalize}_#{bundle_id}.mobileprovision")
```
Same thing here with `.provisionprofile`

joshdholtz commented on fastlane/lib/fastlane/actions/docs/sync_code_signing.md:
```diff
@@ -465,13 +465,13 @@ You'll be asked for the new password on all your machines on the next run.
 ### Import
 
-To import and encrypt a certificate (`.cer`) and the private key (`.p12`) into the _match_ repo run:
+To import and encrypt a certificate (`.cer`), the private key (`.p12`) and the provisioning profiles (`.mobileprovision`) into the _match_ repo run:
```
And here with `.provisionprofile`

joshdholtz commented on fastlane/lib/fastlane/actions/docs/sync_code_signing.md:
````diff
@@ -465,13 +465,13 @@ You'll be asked for the new password on all your machines on the next run.
 ```
 
-You'll be prompted for the certificate (`.cer`) and the private key (`.p12`) paths. _match_ will first validate the certificate (`.cer`) against the Developer Portal before importing the certificate (`.cer`) and the private key (`.p12`).
+You'll be prompted for the certificate (`.cer`), the private key (`.p12`) and the provisioning profiles (`.mobileprovision`) paths. _match_ will first validate the certificate (`.cer`) against the Developer Portal before importing the certificate (`.cer`), the private key (`.p12`) and the provisioning profiles (`.mobileprovision`) into the specified _match_ repository.
````
And here with `.provisionprofile` 😊

joshdholtz reviewed (changes requested):
This looks really good! Just a few comments about supporting `.provisionprofile` as well 😊

DimitarTachev committed:
feat: support macOS provisioning profiles (.provisionprofile) in the `match import` command

DimitarTachev commented:
Hi @joshdholtz,

Thanks for the quick review!

I've addressed the comments for the macOS provisioning profiles support. You could review just the latest commit. I don't have such provisioning profile for a real test but I've added a unit test with a fake profile.

Could you update the PR with a milestone or leave a comment once you have plans for releasing the PR? We will track the PR and specify a minimum Fastlane version for our Windows only flow once we know the exact version 🙃

janpio commented on fastlane/lib/fastlane/actions/docs/sync_code_signing.md:
```diff
@@ -465,13 +465,13 @@ You'll be asked for the new password on all your machines on the next run.
 ### Import
 
-To import and encrypt a certificate (`.cer`) and the private key (`.p12`) into the _match_ repo run:
+To import and encrypt a certificate (`.cer`), the private key (`.p12`) and the provisioning profiles (`.mobileprovision or .provisionprofile`) into the _match_ repo run:
```
```suggestion
To import and encrypt a certificate (`.cer`), the private key (`.p12`) and the provisioning profiles (`.mobileprovision` or `.provisionprofile`) into the _match_ repo run:
```

janpio commented on fastlane/lib/fastlane/actions/docs/sync_code_signing.md:
````diff
@@ -465,13 +465,13 @@ You'll be asked for the new password on all your machines on the next run.
 ```
 
-You'll be prompted for the certificate (`.cer`) and the private key (`.p12`) paths. _match_ will first validate the certificate (`.cer`) against the Developer Portal before importing the certificate (`.cer`) and the private key (`.p12`).
+You'll be prompted for the certificate (`.cer`), the private key (`.p12`) and the provisioning profiles (`.mobileprovision or .provisionprofile`) paths. _match_ will first validate the certificate (`.cer`) against the Developer Portal before importing the certificate, the private key and the provisioning profiles into the specified _match_ repository.
````
```suggestion
You'll be prompted for the certificate (`.cer`), the private key (`.p12`) and the provisioning profiles (`.mobileprovision` or `.provisionprofile`) paths. _match_ will first validate the certificate (`.cer`) against the Developer Portal before importing the certificate, the private key and the provisioning profiles into the specified _match_ repository.
```

Commit:
docs: accept a PR suggestion

Co-Authored-By: Jan Piotrowski <piotrowski+github@gmail.com>

Commit:
docs: accept a PR suggestion

Co-Authored-By: Jan Piotrowski <piotrowski+github@gmail.com>

DimitarTachev commented:
@janpio and @joshdholtz , do you need something else in order to merge this? 🙃

joshdholtz commented:
Getting back to fastlane stuff first thing this morning! Just waking up now 😊 Making some coffee and then I’ll be on but this should be good to merge I believe. I’ll let you know soon 💪

joshdholtz reviewed (approved):
This is so 🔥 Thanks for adding in support for windows and provisioning profiles support into importer 😊 Really appreciate the contribution ❤️ 

(This will be getting released later today)

andrewtheis commented:
@DimitarTachev / @joshdholtz is it possible to bypass validating the certificate with Apple when importing? I have a few projects where I don't have access to the primary account as it's a personal Apple Dev account and not a business account where I can add myself to Certs & Profiles.

joshdholtz commented:
@andrewtheis It "technically" is but you would need to know the ID of the cert 😬 The certificate and private key are named with the ID of the cert. The current process compares the public key content with every certificate on the Apple Developer Portal to find the ID so that it can be named properly in the _match_ repo. 

TBH, I'm not technically sure if that name of the certificate and private key matter now that I think about it 🤔 But it would feel a bit weird to go against what is done in the rest of _match_. 

But if your client only has access to the Certs and Profiles that _should_ be all that is needed I believe. Can you open a new issue with your exact experience and mention me in it? 😇

andrewtheis commented:
So if I specify the team name/ID I should be good?

I don’t have access to the clients Apple ID, they’re making the certs and profiles manually. I previously was importing via some of the manual ruby scripts online that were needed prior to this import addition.

andrewtheis commented:
@Joshdholtz I don’t have access to the client’s Apple ID, they’re making the certs and profiles manually and sending them to me. Using their credentials in fastlane CLI is not an option.

Is there any way to provide the cert ID manually?

I previously was importing via some of the manual ruby scripts online that were needed prior to this import addition. It’s a pain so I was hoping this would work for this use case .

joshdholtz commented:
@andreacipriani I _think_ we can make this possible but it isn't super related to this PR 🙃 Can you create a new issue (feature request) with this in it and mention me? I don't want to lose track of it and you 😊

stefanomondino commented:
@andrewtheis before this feature was actually merged, I did put together this custom action https://github.com/stefanomondino/Jenkins-Murray/blob/master/Jenkins/jenkins/fastlane/actions/custom_import.rb

I basically copy-pasted the normal `import.rb` file and edited where I needed. I'm not a rubyist, so I wasn't able to actually create a proper merge request for the entire fastlane repo (though I would have liked to) but maybe this can be useful.
